conda_mirror/diff_tar.py

```diff
@@ -0,0 +1,209 @@
+"""
+Implementation of the conda-diff-tar command, a tools which allows creating
+differential tarballs of a (usually mirrored) conda repository.  The resulting
+tarball can be used to update a copy of the mirror on a remote (air-gapped)
+system, without having to copy the entire conda repository.
+"""
+import os
+import sys
+import json
+import hashlib
+import tarfile
+from os.path import abspath, isdir, join, relpath
+
+
+MIRROR_DIR = None
+REFERENCE_PATH = './reference.json'
+
+
+def md5_file(path):
+    """
+    Return the MD5 hashsum of the file given by `path` in hexadecimal
+    representation.
+    """
+    h = hashlib.new('md5')
+    with open(path, 'rb') as fi:
+        while 1:
+            chunk = fi.read(262144)
+            if not chunk:
+                break
+            h.update(chunk)
+    return h.hexdigest()
+
+
+def find_repos():
+    """
+    Asssuming the global `MIRROR_DIR` is set, iterate all sub-directories
+    which contain a repodata.json and repodata.json.bz2 file.
+    """
+    for root, unused_dirs, files in os.walk(MIRROR_DIR):
+        if 'repodata.json' in files and 'repodata.json.bz2' in files:
+            yield root
+
+
+def all_repodata():
+    """
+    Return a dictionary mapping all repository sub-directories to the conda
+    package list as respresented by the 'packages' field in repodata.json.
+    """
+    d = {}
+    for repo_path in find_repos():
+        with open(join(repo_path, 'repodata.json')) as fi:
+            index = json.load(fi)['packages']
+        d[repo_path] = index
+    return d
+
+
+def verify_all_repos():
+    """
+    Verify all the MD5 sum of all conda packages listed in all repodata.json
+    files in the repository.
+    """
+    d = all_repodata()
+    for repo_path, index in d.items():
+        for fn, info in index.items():
+            path = join(repo_path, fn)
+            if info['md5'] == md5_file(path):
+                continue
+            print('MD5 mismatch: %s' % path)
+
+
+def write_reference():
+    """
+    Write the "reference file", which is a collection of the content of all
+    repodata.json files.
+    """
+    data = json.dumps(all_repodata(), indent=2, sort_keys=True)
+    # make sure we have newline at the end
+    if not data.endswith('\n'):
+        data += '\n'
+    with open(REFERENCE_PATH, 'w') as fo:
+        fo.write(data)
+
+
+def read_reference():
+    """
+    Read the "reference file" from disk and return its content as a dictionary.
+    """
+    with open(REFERENCE_PATH) as fi:
+        return json.load(fi)
+
+
+def get_updates():
+    """
+    Compare the "reference file" to the actual the repository (all the
+    repodata.json files) and iterate the new and updates files in the
+    repository.  That is, the files which need to go into the differential
+    tarball.
+    """
+    try:
+        d1 = read_reference()
```

Comment:
Please `os.exists` instead of exception-based programming. You can raise the `sys.exit` directly in `read_reference`

Reply:
I guess you mean `os.path.exists`.  I'll call `sys.exit` in the function directly again, based on the exception because "asking for forgiveness is better than asking for permission" :-)
b260c5bbd65d7408e46f95882434807f774f8c48